src/main/java/com/direwolf20/buildinggadgets/common/util/blocks/RegionSnapshot.java

```diff
@@ -0,0 +1,247 @@
+        palettes.add(new NBTTagCompound());
+        // Palette serialization end
+
+        // Streaking: a way to record a number of repeating block states, using only 2 integers
+        // Storage frame: one or two integers indicating a sequence of block states
+        // Regular frame: a type of storage frame; 1 integer, representing the ID of the block state recorded
+        // Streak frame: a type of storage frame; 2 integers, where the first one is positive, representing the ID of the repeating block state; the second one is negative, representing the number of repeating block states
+
+        // Complete storage frames, will be serialized into a NBTTagList later
+        IntList frames = new IntArrayList();
+
+        // The repeating block state, if any
+        IBlockState lastStreak = null;
+        // Number of repeating block states
+        int streak = 0;
+
+        boolean firstIteration = true;
+        for (IBlockState state : snapshot.blockStates) {
+            // Palette serialization begin
+            if (!recordedPalettes.contains(state)) {
+                recordedPalettes.add(state);
+                // ID 0 is preoccupied for "nothing should be placed here", which is represented in 'null'
+                mapPalettes.put(state, mapPalettes.size() + 1);
+                palettes.add(NBTUtil.writeBlockState(state));
+            }
+            // Palette serialization ends
+
+            // Hence we record the storage frame after it is completed, the inital values of the counter would be used as a storage frame regardless, therefore we need to skip inorder for this algorithm to function correctly
+            if (firstIteration)
+                firstIteration = false;
+            else {
+                // If the same block state showed up again, increase the number of streaks
+                if (state == lastStreak) {
+                    streak++;
+                    continue;
+                }
+
+                // Note: if the program reached here, it means the current storage frame is completed
+
+                // If it didn't show up and this block is not by itself, record the streak frame
+                // The first one represents the ID of the streaking block state, the second one, which is negative, represents the amount of streaking block states
+                if (streak != 0) {
+                    // ID of the block state that is streaking
+                    frames.add(mapPalettes.getInt(lastStreak));
+                    // Number of streaks
+                    // Stored in negative to differentiate with a block state ID
+                    frames.add(-streak);
+                } else
+                    // Otherwise just add the block state ID, since there is only one of it
+                    frames.add(mapPalettes.getInt(lastStreak));
+            }
+
+            // Regardless of which situation, we reset the counter
+            streak = 0;
```

Comment:
That should be `1`
```suggestion
            streak = 1;
```

Reply:
I'd leave it at 0, and make 0 indicate no streak

Reply:
This is actually a reminder of myself, that this thing won't work if we reset the counter to 0. 

Here is an example why it doesn't work
```
// Each letter is a block
|x|x|y|
 ^ streak = 0

|x|x|y|
   ^ streak = 1

|x|x|y|
     ^ record (x, 1), reset streak = 0 
// It should record (x, 2)
```

Reply:
It could also record (x, 1) that does not make a difference, if your deserilisation just always places one nore then the streak's size... 
This would actually be an advantage of using 0 - there's no need to differ between streak and no streak.

Reply:
With your addition, there is no need to make a difference at all - regardless of which counter initial we use.